File: day01/go/anoff/solution.go
package main

import (
	"fmt"
	"math"
)

// tag::part1[]
func main() {
	input := readInput("./input.txt")
	part1(StringSlice2IntSlice(input))
}

func part1(massInventory []int) {
	var fuelSum int = 0
	for _, mass := range massInventory {
		fmt.Println(mass)
		fuelSum += RequiredFuel(mass)
	}
	fmt.Println("Solution for part1:", fuelSum)
}

// end::part1[]
func part2() {

}

// tag::part1[]
func RequiredFuel(mass int) int {
	var massFloat = float64(mass)
	var fuelFloat = math.Floor(massFloat / 3)
	return int(fuelFloat) - 2
}

// end::part1[]

File: day01/ruby/anoff/solution.rb
#!/usr/bin/ruby
def required_fuel(mass)
  (mass / 3) - 2
end

def part1(massInventory)
  fuelSum = massInventory
    .map { |mass| required_fuel(mass) }
    .reduce(0) { |sum, num| sum + num }
end

def part2(massInventory)
  fuelSum = 0
  massInventory.each{ |mass|
    fuel = required_fuel(mass)
    while fuel > 0 do
      fuelSum += fuel
      fuel = required_fuel(fuel)
    end
  }
  fuelSum
end

if caller.length == 0
  input = File.read("./input.txt")

  massInventory = input.split("\n").map(&:to_i)

  puts "Solution for part1: %d" % part1(massInventory)
  puts "Solution for part2: %d" % part2(massInventory)
end